examples/door-lock-app/linux/src/LockEndpoint.cpp

```diff
@@ -0,0 +1,339 @@
+    user.createdBy      = userInDb.createdBy;
+    user.lastModifiedBy = userInDb.lastModifiedBy;
+
+    ChipLogDetail(
+        Zcl,
+        "Found occupied user "
+        "[endpoint=%d,adjustedIndex=%hu,name=\"%*.s\",credentialsCount=%zu,uniqueId=%x,type=%" PRIu8 ",credentialRule=%" PRIu8 ","
+        "createdBy=%d,lastModifiedBy=%d]",
+        mEndpointId, adjustedUserIndex, static_cast<int>(user.userName.size()), user.userName.data(), user.credentials.size(),
+        user.userUniqueId, to_underlying(user.userType), to_underlying(user.credentialRule), user.createdBy, user.lastModifiedBy);
+
+    return true;
+}
+
+bool LockEndpoint::SetUser(uint16_t userIndex, chip::FabricIndex creator, chip::FabricIndex modifier,
+                           const chip::CharSpan & userName, uint32_t uniqueId, DlUserStatus userStatus, DlUserType usertype,
+                           DlCredentialRule credentialRule, const DlCredential * credentials, size_t totalCredentials)
+{
+    ChipLogProgress(Zcl,
+                    "Door Lock App: LockEndpoint::SetUser "
+                    "[endpoint=%d,userIndex=%" PRIu16 ",creator=%d,modifier=%d,userName=\"%*.s\",uniqueId=%" PRIx32
+                    ",userStatus=%" PRIu8 ",userType=%" PRIu8 ","
+                    "credentialRule=%" PRIu8 ",credentials=%p,totalCredentials=%zu]",
+                    mEndpointId, userIndex, creator, modifier, static_cast<int>(userName.size()), userName.data(), uniqueId,
+                    to_underlying(userStatus), to_underlying(usertype), to_underlying(credentialRule), credentials,
+                    totalCredentials);
+
+    uint16_t adjustedUserIndex = static_cast<uint16_t>(userIndex - 1);
+    if (adjustedUserIndex > mLockUsers.size())
+    {
+        ChipLogError(Zcl, "Cannot set user - index out of range [endpoint=%d,index=%d,adjustedUserIndex=%" PRIu16 "]", mEndpointId,
+                     userIndex, adjustedUserIndex);
+        return false;
+    }
+
+    auto & userInStorage = mLockUsers[adjustedUserIndex];
+
+    if (userName.size() > DOOR_LOCK_MAX_USER_NAME_SIZE)
+    {
+        ChipLogError(Zcl, "Cannot set user - user name is too long [endpoint=%d,index=%d,adjustedUserIndex=%" PRIu16 "]",
+                     mEndpointId, userIndex, adjustedUserIndex);
+        return false;
+    }
+
+    strncpy(userInStorage.userName, userName.data(), userName.size());
+    userInStorage.userName[userName.size()] = 0;
+    userInStorage.userUniqueId              = uniqueId;
+    userInStorage.userStatus                = userStatus;
+    userInStorage.userType                  = usertype;
+    userInStorage.credentialRule            = credentialRule;
+    userInStorage.lastModifiedBy            = modifier;
+    userInStorage.createdBy                 = creator;
+
+    userInStorage.totalCredentials = totalCredentials;
+    for (size_t i = 0; i < totalCredentials; ++i)
+    {
+        userInStorage.credentials[i] = credentials[i];
```

Comment:
What guarantees that we don't write past the end of the `userInStorage.credentials` array?

Reply:
It is implicitly checked in the [cluster code itself](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/door-lock-server/door-lock-server.cpp#L1447) but I've added explicit check here too.